docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+to introduce a special response code for `CheckTx` to indicate this behaviour, and ban the peers
+who gossip such transactions. Additionally, users expressed a need for banning peers who
+repeadetly send transactions failing `CheckTx`. 
+
+The main goal of this document is to analyse the cases where peers could be banned when they send 
+transacations failing `CheckTx`, and provide the exact conditions that a peer and transaction have
+to satisfy in order to mark the peer as bad.
+
+This document will also include a proposal for implementing these changes within the mempool, including
+potential changes to the existing mempool logic and implementation.
+
+
+## Background
+
+This work was triggered by issue [#7918](https://github.com/tendermint/tendermint/issues/7918) and a related
+discussion in [#2185](https://github.com/tendermint/tendermint/issues/2185). Additionally,
+there was a [proposal](https://github.com/tendermint/tendermint/issues/6523) 
+to disconnect from peers after they send us transactions that constantly fail `CheckTx`.  While 
+the actual implementation of an additional response code for `CheckTx` is straight forward there 
+are certain things to consider. The questions to answer, along with identified risks will be outlined in 
+the discussion. 
+
+### Existing issues and concerns
+
+Before diving into the details, we collected a set of issues opened by various users, arguing for 
+this behaviour and explaining their needs. 
+
+- Celestia: [blacklisting peers that repeatedly send bad tx](https://github.com/celestiaorg/celestia-core/issues/867)
+  and investigating how [Tendermint treats large Txs](https://github.com/celestiaorg/celestia-core/issues/243)
+- BigChainDb: [Handling proposers who propose bad blocks](https://github.com/bigchaindb/BEPs/issues/84)
+- IBC relayer: Simulate transactions in the mempool and detect whether they could ever have been valid to 
+  [avoid overloading the p2p/mempool system](https://github.com/cosmos/ibc-go/issues/853#issuecomment-1032211020)
+ (ToDo check with Adi what is this about, what are valid reasons this can happen and when should we disconnect)
+  (relevant code [here](https://github.com/cosmos/ibc-go/blob/a0e59b8e7a2e1305b7b168962e20516ca8c98fad/modules/core/ante/ante.go#L23)) 
+- From  Adi: Nodes allow transactions with a wrong `minGas` transaction and gossip them, and other nodes keep rejecting them. (Problem seen with relayers)
+
+### Current state of mempool/p2p interaction
+
+Transactions received from a peer are handled within the [`Receive`](https://github.com/tendermint/tendermint/blob/ff0f98892f24aac11e46aeff2b6d2c0ad816701a/mempool/v0/reactor.go#L158) routine. 
+
+Currently, the mempool triggers a disconnect from a peer in the case of the following errors:
+
+  - [Unknown message type](https://github.com/tendermint/tendermint/blob/ff0f98892f24aac11e46aeff2b6d2c0ad816701a/mempool/v0/reactor.go#L184)
```

Comment:
Question for those working on the p2p spec.
My understanding of the spec is that the "transport" part of p2p negotiates what channels each node in the pair supports. So, wouldn't receiving an unknown message also be a reason for banning?

Reply:
This error happens when a message that is not a transaction gets send via the right channel.